## Machine Learning

This notebook summarizes the performance of various different machine learning models when applied to the dataset.

The first step I will take is to import some basic python packages and load the dataset into the `df` variable.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.options.display.max_columns = 500

%matplotlib inline

In [2]:
df = pd.read_csv('df_model_5.csv', index_col=0)

In [5]:
df.head()

casename        date     id  \
0  ACT-UP Triangle v. Commission for Health Services  1997-04-11  53839   
1                   Mahoney v. Ronnie’s Road Service  1997-03-07  53841   
2                                State v. Westbrooks  1996-12-06  53843   
3                                    State v. Conner  1997-02-10  53847   
4                           Fulton Corp. v. Faulkner  1997-02-10  53848   

          judge                                            opinion      type  \
0          frye  FRYE, Justice.\nThis case involves the adoptio...  majority   
1    per_curiam  PER CURIAM.\nAFFIRMED.\nJustice PARKER did not...  majority   
2  parker_sarah  PARKER, Justice.\nDefendant, Donna Sue Westbro...  majority   
3      whichard  WHICHARD, Justice.\nOn 13 November 1990, defen...  majority   
4          webb  WEBB, Justice.\nThis case brings to the Court ...  majority   

   judge_count  year  word_count  sentence_count  avg_sent_length  polarity  \
0          415  1997        4364             214        20.392523  0.102371   
1         6440  1997          16               3         5.333333  0.000000   
2          456  1996        9467             405        23.375309 -0.012108   
3          661  1997        6111             268        22.802239 -0.024927   
4          801  1997        1346              75        17.946667  0.051909   

   subjectivity  
0      0.473721  
1      0.000000  
2      0.321622  
3      0.415121  
4      0.415155

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80749 entries, 0 to 80748
Data columns (total 13 columns):
casename           80749 non-null object
date               80748 non-null object
id                 80749 non-null int64
judge              80749 non-null object
opinion            80749 non-null object
type               80749 non-null object
judge_count        80749 non-null int64
year               80749 non-null int64
word_count         80749 non-null int64
sentence_count     80749 non-null int64
avg_sent_length    80749 non-null float64
polarity           80749 non-null float64
subjectivity       80749 non-null float64
dtypes: float64(3), int64(5), object(5)
memory usage: 8.6+ MB


## Preparing the Features for Modeling

I will start by testing how some classifiers perform using only the characteristic features of the opinions that I used when exploring the data without the actual text of the opinions.  In order to do this, I need to convert those features into an appropriate format for the machine learning models.

### Creating the Machine Learning Dataframe

I will begin by creating a separate dataframe, `df_ml`, where I will store the various features that we will use in training the models.  I will start by adding the id number for each opinion for identification purposes, as well as the judge, which will be our label that the model will attempt to predict.

In [6]:
df_ml = df.loc[:, ['id', 'judge']]
df_ml.head()

id         judge
0  53839          frye
1  53841    per_curiam
2  53843  parker_sarah
3  53847      whichard
4  53848          webb

### Converting the Categorical Features into Dummy Variables

The dataset has two categorical features, the type of opinion (majority, concurrence, etc.) and the year the opinion was issued.  I will convert both of these features into dummy columns using `pd.get_dummies` and then add the dummies to the `df_ml` dataframe:

In [7]:
# create dummy columns for 'type' feature
df_dummies = pd.get_dummies(df['type'], prefix='type')
df_dummies.head()

type_concurrence  type_concurring-in-part-and-dissenting-in-part  \
0                 0                                               0   
1                 0                                               0   
2                 0                                               0   
3                 0                                               0   
4                 0                                               0   

   type_dissent  type_majority  type_rehearing  
0             0              1               0  
1             0              1               0  
2             0              1               0  
3             0              1               0  
4             0              1               0

In [9]:
# add the dummy columns to the df_ml dataframe
df_ml = pd.concat([df_ml, df_dummies], axis=1)
df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  
0              1               0  
1              1               0  
2              1               0  
3              1               0  
4              1               0

In [10]:
# create dummy columns for 'year' feature
df_dummies = pd.get_dummies(df['year'], prefix='year')
df_dummies.head()

year_1779  year_1784  year_1787  year_1789  year_1790  year_1791  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1792  year_1793  year_1794  year_1795  year_1796  year_1797  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1798  year_1799  year_1800  year_1801  year_1802  year_1803  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1804  year_1805  year_1806  year_1807  year_1808  year_1809  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1810  year_1811  year_1812  year_1813  year_1814  year_1815  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1816  year_1817  year_1818  year_1819  year_1820  year_1821  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1822  year_1823  year_1824  year_1825  year_1826  year_1827  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1828  year_1829  year_1830  year_1831  year_1832  year_1833  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1834  year_1835  year_1836  year_1837  year_1838  year_1839  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1840  year_1841  year_1842  year_1843  year_1844  year_1845  \
0          0          0          0          0          0          0   
1          0      

In [11]:
# add the dummy columns to the df_ml dataframe
df_ml = pd.concat([df_ml, df_dummies], axis=1)
df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  year_1779  year_1784  year_1787  year_1789  \
0              1               0          0          0          0          0   
1              1               0          0          0          0          0   
2              1               0          0          0          0          0   
3              1               0          0          0          0          0   
4              1               0          0          0          0          0   

   year_1790  year_1791  year_1792  year_1793  year_1794  year_1795  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1796  year_1797  year_1798  year_1799  year_1800  year_1801  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1802  year_1803  year_1804  year_1805  year_1806  year_1807  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1808  year_1809  year_1810  year_1811  year_1812  year_1813  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1814  year_1815  year_1816  year_1817  year_1818  year_1819  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1820  year_1821  year_1822  year_1823  year_1824  year_1825  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1826  year_1827  year_1828  year_1829  year_1830  year_1831  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          

### Normalize the Continuous Variables

The dataset has a number of continuous variables that I need to normalize, including the word count, sentence count, and average sentence lenghth.  Also, the polarity scores from `TextBlob` are on a -1-to-1 scale, so I will convert those values into a normalized 0-to-1 range.

In [15]:
# Import and instantiate MinMaxScaler
from sklearn.preprocessing import MinMaxScaler
min_max_scaler = MinMaxScaler()

# Normalize the relevant features
x = df.iloc[:, 8:12].values
x_scaled = min_max_scaler.fit_transform(x)

# Save the normalized values into a dataframe
df_scaled = pd.DataFrame(x_scaled, columns=['word_count', 'sentence_count', 'avg_sent_length', 'polarity'])
df_scaled.head()

word_count  sentence_count  avg_sent_length  polarity
0    0.088804        0.098247         0.093527  0.501317
1    0.000265        0.000923         0.018977  0.444444
2    0.192718        0.186347         0.108294  0.437718
3    0.124379        0.123155         0.105457  0.430596
4    0.027348        0.034133         0.081419  0.473283

In [17]:
# add the normalized features to the df_ml dataframe
df_ml = pd.concat([df_ml, df_scaled], axis=1)

# add the already-normalized subjectivity column to the dl_ml dataframe
df_ml['subjectivity'] = df.loc[:, 'subjectivity']

df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  year_1779  year_1784  year_1787  year_1789  \
0              1               0          0          0          0          0   
1              1               0          0          0          0          0   
2              1               0          0          0          0          0   
3              1               0          0          0          0          0   
4              1               0          0          0          0          0   

   year_1790  year_1791  year_1792  year_1793  year_1794  year_1795  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1796  year_1797  year_1798  year_1799  year_1800  year_1801  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1802  year_1803  year_1804  year_1805  year_1806  year_1807  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1808  year_1809  year_1810  year_1811  year_1812  year_1813  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1814  year_1815  year_1816  year_1817  year_1818  year_1819  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1820  year_1821  year_1822  year_1823  year_1824  year_1825  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1826  year_1827  year_1828  year_1829  year_1830  year_1831  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          

In [18]:
# save the df_ml dataframe to the local drive
export_df = df_ml.to_csv('df_ml_1.csv')

## Training Some Benchmark Models

In [36]:
df_ml2 = df_ml[df_ml['judge'] != 'per_curiam']
df_ml2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74309 entries, 0 to 80747
Columns: 247 entries, id to subjectivity
dtypes: float64(9), int64(1), object(1), uint8(236)
memory usage: 23.5+ MB


In [45]:
# divide the data into label and features for use in ml models
X = df_ml.iloc[:, 2:]
y = df_ml.loc[:, 'judge']
X.columns

Index(['type_concurrence', 'type_concurring-in-part-and-dissenting-in-part',
       'type_dissent', 'type_majority', 'type_rehearing', 'year_1779',
       'year_1784', 'year_1787', 'year_1789', 'year_1790',
       ...
       'year_2017', 'word_count', 'sentence_count', 'avg_sent_length',
       'polarity', 'word_count', 'sentence_count', 'avg_sent_length',
       'polarity', 'subjectivity'],
      dtype='object', length=245)

In [47]:
# split the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=6, stratify=y)

### Logistic Regression

In [48]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=6)
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
print('Accuracy on training set = {}'.format(lr.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(lr.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy on training set = 0.28983999603705357
Accuracy on test set = 0.2698632851198732


### Stochastic Gradient Descent Classifier

In [49]:
from sklearn.linear_model import SGDClassifier
sgdc = SGDClassifier(random_state=6)
sgdc.fit(X_train, y_train)
y_pred_sgdc = sgdc.predict(X_test)
print('Accuracy on training set = {}'.format(sgdc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(sgdc.score(X_test, y_test)))

Accuracy on training set = 0.23133699905880023
Accuracy on test set = 0.22359817713493163


### Linear Support Vector Machine Classifier

In [50]:
from sklearn.svm import LinearSVC
svc = LinearSVC(random_state=6)
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)
print('Accuracy on training set = {}'.format(svc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(svc.score(X_test, y_test)))

Accuracy on training set = 0.302174666864814
Accuracy on test set = 0.2855656825837131


### Multinomial Naive Bayes

In [51]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()
nb.fit(X_train, y_train)
y_pred_nb = nb.predict(X_test)
print('Accuracy on training set = {}'.format(nb.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(nb.score(X_test, y_test)))

Accuracy on training set = 0.2104324565314311
Accuracy on test set = 0.20001981375074301


### Random Forest Classifier

In [52]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=6)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)
print('Accuracy on training set = {}'.format(rf.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(rf.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


Accuracy on training set = 0.9885404798467661
Accuracy on test set = 0.24148008718050326


### Summary

Here is a summary of the results:

| Model               | Train  | Test   |
|---------------------|--------|--------|
| Linear SVM          | 0.3022 | 0.2856 |
| Logistic Regression | 0.2898 | 0.2699 |
| Random Forest       | 0.9885 | 0.2415 |
| SGD                 | 0.2313 | 0.2234 |
| Naive Bayes         | 0.2104 | 0.2000 |

The best performing models were able to correctly predict the judge who wrote the opinion just over 25% of the time on the test set.  The top three performing models also appear to have overfit on the training set (the Random Forest Classifier to a much more significant degree).  While this is far from our ultimate goal, it does demonstrate that these features contain some signal.

Next, I will use `TfidfVectorizer` to vectorize the text of the opinions into additional features to see if it improves model performance.

## Vectorized Text Features

As discussed in the Data Preprocessing notebook, I have already used `TfidfVectorizer` on the corpus with varying `min_df` values to build vocabularies.  I will read in the vectorized text files and see how the models perform.  I will first test the smallest vocabulary of 3,677 words/features. 

In [3]:
# read in the saved tfidf vectors from local .csv file
df_tfidf = pd.read_csv('df_tfidf_02.csv', index_col=0)

In [5]:
df_tfidf.head()

00       000        10       100  101  102       103  104       105  \
0  0.000000  0.000000  0.000000  0.000000  0.0  0.0  0.000000  0.0  0.000000   
1  0.000000  0.000000  0.000000  0.000000  0.0  0.0  0.000000  0.0  0.000000   
2  0.002442  0.017602  0.005822  0.000000  0.0  0.0  0.003096  0.0  0.000000   
3  0.000000  0.004179  0.009676  0.004278  0.0  0.0  0.000000  0.0  0.000000   
4  0.000000  0.000000  0.000000  0.000000  0.0  0.0  0.000000  0.0  0.119991   

   106  107       108  109        11       110  111       112  113       114  \
0  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0  0.006115  0.0  0.017759   
1  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0  0.000000  0.0  0.000000   
2  0.0  0.0  0.000000  0.0  0.002189  0.000000  0.0  0.000000  0.0  0.000000   
3  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0  0.000000  0.0  0.000000   
4  0.0  0.0  0.014679  0.0  0.000000  0.015228  0.0  0.000000  0.0  0.000000   

        115  116       117       118      119        12  120  121       122  \
0  0.000000  0.0  0.000000  0.017337  0.00000  0.008272  0.0  0.0  0.011754   
1  0.000000  0.0  0.000000  0.000000  0.00000  0.000000  0.0  0.0  0.000000   
2  0.000000  0.0  0.002969  0.000000  0.00304  0.002079  0.0  0.0  0.000000   
3  0.005137  0.0  0.000000  0.000000  0.00000  0.000000  0.0  0.0  0.009819   
4  0.000000  0.0  0.015045  0.000000  0.00000  0.021065  0.0  0.0  0.000000   

        123      124       125  126  127       128  129        13       130  \
0  0.006137  0.01231  0.000000  0.0  0.0  0.000000  0.0  0.000000  0.000000   
1  0.000000  0.00000  0.000000  0.0  0.0  0.000000  0.0  0.000000  0.000000   
2  0.003085  0.00000  0.000000  0.0  0.0  0.000000  0.0  0.027079  0.000000   
3  0.000000  0.00000  0.000000  0.0  0.0  0.004998  0.0  0.003750  0.014608   
4  0.000000  0.00000  0.030301  0.0  0.0  0.000000  0.0  0.000000  0.000000   

        131       132       133      134       135       136  137  138  139  \
0  0.000000  0.000000  0.000000  0.00000  0.006142  0.000000  0.0  0.0  0.0   
1  0.000000  0.000000  0.000000  0.00000  0.000000  0.000000  0.0  0.0  0.0   
2  0.003006  0.000000  0.000000  0.00000  0.006175  0.002944  0.0  0.0  0.0   
3  0.009991  0.004744  0.004931  0.02001  0.005131  0.004892  0.0  0.0  0.0   
4  0.000000  0.000000  0.015031  0.00000  0.000000  0.000000  0.0  0.0  0.0   

         14  140  141       142       143       144  145       146       147  \
0  0.000000  0.0  0.0  0.000000  0.018073  0.005839  0.0  0.000000  0.000000   
1  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.0  0.000000  0.000000   
2  0.004202  0.0  0.0  0.000000  0.000000  0.000000  0.0  0.000000  0.000000   
3  0.003492  0.0  0.0  0.005029  0.005032  0.000000  0.0  0.005079  0.005121   
4  0.000000  0.0  0.0  0.000000  0.000000  0.000000  0.0  0.000000  0.000000   

   148  149        15  150  151       152       153  154  155  156       157  \
0  0.0  0.0  0.004249  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000000   
1  0.0  0.0  0.000000  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000000   
2  0.0  0.0  0.008542  0.0  0.0  0.003063  0.002971  0.0  0.0  0.0  0.003086   
3  0.0  0.0  0.003549  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.005129   
4  0.0  0.0  0.000000  0.0  0.0  0.000000  0.000000  0.0  0.0  0.0  0.000000   

       158  159       15a        16       160       161  162       163  164  \
0  0.00000  0.0  0.052304  0.004528  0.000000  0.005937  0.0  0.000000  0.0   
1  0.00000  0.0  0.000000  0.000000  0.000000  0.000000  0.0  0.000000  0.0   
2  0.00000  0.0  0.014606  0.027311  0.002995  0.000000  0.0  0.000000  0.0   
3  0.00488  0.0  0.072821  0.000000  0.000000  0.000000  0.0  0.005014  0.0   
4  0.00000  0.0  0.000000  0.000000  0.000000  0.000000  0.0  0.015285  0.0   

       165  166  167       168  169        17       170  171  172  173  \
0  0.00617  0.0  0.0  0.000000  0.0  0.004638  0.000000  0.0  0.0  0.0   
1  0.00000  0.0  0.0  0.000000  0.0  

In [4]:
# create the machine learing dataframe
df_ml = df.loc[:, ['id', 'judge']]
df_ml.head()

id         judge
0  53839          frye
1  53841    per_curiam
2  53843  parker_sarah
3  53847      whichard
4  53848          webb

In [7]:
# add the vectorized vocabulary to the df_ml dataframe
df_ml = pd.concat([df_ml, df_tfidf], axis=1)

In [8]:
df_ml.head()

id         judge        00       000        10       100  101  102  \
0  53839          frye  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
1  53841    per_curiam  0.000000  0.000000  0.000000  0.000000  0.0  0.0   
2  53843  parker_sarah  0.002442  0.017602  0.005822  0.000000  0.0  0.0   
3  53847      whichard  0.000000  0.004179  0.009676  0.004278  0.0  0.0   
4  53848          webb  0.000000  0.000000  0.000000  0.000000  0.0  0.0   

        103  104       105  106  107       108  109        11       110  111  \
0  0.000000  0.0  0.000000  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0   
1  0.000000  0.0  0.000000  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0   
2  0.003096  0.0  0.000000  0.0  0.0  0.000000  0.0  0.002189  0.000000  0.0   
3  0.000000  0.0  0.000000  0.0  0.0  0.000000  0.0  0.000000  0.000000  0.0   
4  0.000000  0.0  0.119991  0.0  0.0  0.014679  0.0  0.000000  0.015228  0.0   

        112  113       114       115  116       117       118      119  \
0  0.006115  0.0  0.017759  0.000000  0.0  0.000000  0.017337  0.00000   
1  0.000000  0.0  0.000000  0.000000  0.0  0.000000  0.000000  0.00000   
2  0.000000  0.0  0.000000  0.000000  0.0  0.002969  0.000000  0.00304   
3  0.000000  0.0  0.000000  0.005137  0.0  0.000000  0.000000  0.00000   
4  0.000000  0.0  0.000000  0.000000  0.0  0.015045  0.000000  0.00000   

         12  120  121       122       123      124       125  126  127  \
0  0.008272  0.0  0.0  0.011754  0.006137  0.01231  0.000000  0.0  0.0   
1  0.000000  0.0  0.0  0.000000  0.000000  0.00000  0.000000  0.0  0.0   
2  0.002079  0.0  0.0  0.000000  0.003085  0.00000  0.000000  0.0  0.0   
3  0.000000  0.0  0.0  0.009819  0.000000  0.00000  0.000000  0.0  0.0   
4  0.021065  0.0  0.0  0.000000  0.000000  0.00000  0.030301  0.0  0.0   

        128  129        13       130       131       132       133      134  \
0  0.000000  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.00000   
1  0.000000  0.0  0.000000  0.000000  0.000000  0.000000  0.000000  0.00000   
2  0.000000  0.0  0.027079  0.000000  0.003006  0.000000  0.000000  0.00000   
3  0.004998  0.0  0.003750  0.014608  0.009991  0.004744  0.004931  0.02001   
4  0.000000  0.0  0.000000  0.000000  0.000000  0.000000  0.015031  0.00000   

        135       136  137  138  139        14  140  141       142       143  \
0  0.006142  0.000000  0.0  0.0  0.0  0.000000  0.0  0.0  0.000000  0.018073   
1  0.000000  0.000000  0.0  0.0  0.0  0.000000  0.0  0.0  0.000000  0.000000   
2  0.006175  0.002944  0.0  0.0  0.0  0.004202  0.0  0.0  0.000000  0.000000   
3  0.005131  0.004892  0.0  0.0  0.0  0.003492  0.0  0.0  0.005029  0.005032   
4  0.000000  0.000000  0.0  0.0  0.0  0.000000  0.0  0.0  0.000000  0.000000   

        144  145       146       147  148  149        15  150  151       152  \
0  0.005839  0.0  0.000000  0.000000  0.0  0.0  0.004249  0.0  0.0  0.000000   
1  0.000000  0.0  0.000000  0.000000  0.0  0.0  0.000000  0.0  0.0  0.000000   
2  0.000000  0.0  0.000000  0.000000  0.0  0.0  0.008542  0.0  0.0  0.003063   
3  0.000000  0.0  0.005079  0.005121  0.0  0.0  0.003549  0.0  0.0  0.000000   
4  0.000000  0.0  0.000000  0.000000  0.0  0.0  0.000000  0.0  0.0  0.000000   

        153  154  155  156       157      158  159       15a        16  \
0  0.000000  0.0  0.0  0.0  0.000000  0.00000  0.0  0.052304  0.004528   
1  0.000000  0.0  0.0  0.0  0.000000  0.00000  0.0  0.000000  0.000000   
2  0.002971  0.0  0.0  0.0  0.003086  0.00000  0.0  0.014606  0.027311   
3  0.000000  0.0  0.0  0.0  0.005129  0.00488  0.0  0.072821  0.000000   
4  0.000000  0.0  0.0  0.0  0.000000  0.00000  0.0  0.000000  0.000000   

        160       161  162       163  164      165  166  167       168  169  \
0  0.000000  0.005937  0.0  0.000000  0.0  0.00617  0.0  0.0  0.000000  0.0   
1  0.000000  0.000000  0.0  0.000000  0.0  0.00000  0.0  0.0  0.000000  0.0   
2  0.002995  0.000000  0.0  0.000000  0.0  0.00000  0.0  0.

In [9]:
# divide the data into label and features for use in ml models
X = df_ml.iloc[:, 2:]
y = df_ml.loc[:, 'judge']
X.columns

Index(['00', '000', '10', '100', '101', '102', '103', '104', '105', '106',
       ...
       'wrongfully', 'wrote', 'year', 'years', 'yes', 'yet', 'york', 'you',
       'young', 'your'],
      dtype='object', length=3677)

In [11]:
# split the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=6, stratify=y)

### Logistic Regression

In [12]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=6)
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
print('Accuracy on training set = {}'.format(lr.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(lr.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy on training set = 0.581562391638183
Accuracy on test set = 0.49217356845650884


### Stochastic Gradient Descent Classifier

In [13]:
from sklearn.linear_model import SGDClassifier
sgdc = SGDClassifier(random_state=6)
sgdc.fit(X_train, y_train)
y_pred_sgdc = sgdc.predict(X_test)
print('Accuracy on training set = {}'.format(sgdc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(sgdc.score(X_test, y_test)))

Accuracy on training set = 0.7290335364343389
Accuracy on test set = 0.5581038240538934


### Linear Support Vector Machine Classifier

In [14]:
from sklearn.svm import LinearSVC
svc = LinearSVC(random_state=6)
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)
print('Accuracy on training set = {}'.format(svc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(svc.score(X_test, y_test)))

Accuracy on training set = 0.9444361883060055
Accuracy on test set = 0.6814444224291658


### Multinomial Naive Bayes

In [16]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()
nb.fit(X_train, y_train)
y_pred_nb = nb.predict(X_test)
print('Accuracy on training set = {}'.format(nb.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(nb.score(X_test, y_test)))

Accuracy on training set = 0.23239378477898318
Accuracy on test set = 0.21111551416683177


### Random Forest Classifier

In [19]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=900, max_depth=9, min_samples_split=600, min_samples_leaf=50, random_state=6)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)
print('Accuracy on training set = {}'.format(rf.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(rf.score(X_test, y_test)))

Accuracy on training set = 0.24644573240204093
Accuracy on test set = 0.22770953041410738


### Summary

Here is a summary of the results:

| Model (min_df=0.02) | Train  | Test   |
|---------------------|--------|--------|
| Linear SVM          | 0.9044 | 0.6814 |
| Logistic Regression | 0.5816 | 0.4922 |
| SGD                 | 0.7290 | 0.5581 |
| Random Forest       | 0.2464 | 0.2277 |
| Naive Bayes         | 0.2324 | 0.2111 |

Three of the classifiers performed substantially better with the vectorized text.  I will drop the Random Forest and Naive Bayes models given their inferior performance.  Let's see how the Linear SVM, Logistic Regression, and SGD models do with an expanded vocabulary of 16905 words/features:

In [3]:
# create stop words for use in vectorizer
stop_words = list(set(df['judge']))
add_stop_words = ['clark', 'johnson', 'martin', 'parker', 'timmons', 'goodson', 'walker', 'judge', 'justice', 'per', 'curiam']
stop_words += add_stop_words

In [4]:
# vectorize the corpus using TfidfVectorizer and a min_df of 0.001
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(stop_words=stop_words, min_df=0.001)
data_tfidf = tfidf.fit_transform(df['opinion'])
df_tfidf = pd.DataFrame(data_tfidf.toarray(), columns=tfidf.get_feature_names())
df_tfidf.index = df.index

df_tfidf.head()

00       000   01        02        03   04        05  050   06   07  \
0  0.000000  0.000000  0.0  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0   
1  0.000000  0.000000  0.0  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0   
2  0.002282  0.016448  0.0  0.000000  0.000000  0.0  0.000000  0.0  0.0  0.0   
3  0.000000  0.003825  0.0  0.000000  0.006578  0.0  0.006501  0.0  0.0  0.0   
4  0.000000  0.000000  0.0  0.020194  0.000000  0.0  0.019954  0.0  0.0  0.0   

    08        09   0f        10       100  1000  1001  1002  1003  1004  1005  \
0  0.0  0.000000  0.0  0.000000  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   
1  0.0  0.000000  0.0  0.000000  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   
2  0.0  0.000000  0.0  0.005440  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   
3  0.0  0.000000  0.0  0.008856  0.003916   0.0   0.0   0.0   0.0   0.0   0.0   
4  0.0  0.020754  0.0  0.000000  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   

   1006  1007  1008  1009  101  1010  1011  1012  1013  1014  1015  1016  \
0   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4   0.0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   1017  1018  1019  102  1020  1021  1022  1023  1024  1025  1026  1027  \
0   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4   0.0   0.0   0.0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   1028  1029       103  1030  1031  1032  1033  1035  1036  1037  1038  1039  \
0   0.0   0.0  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1   0.0   0.0  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2   0.0   0.0  0.002893   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
3   0.0   0.0  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4   0.0   0.0  0.000000   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   

   104  1040  1041  1042  1043  1044  1045  1046  1047  1048  1049       105  \
0  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.000000   
1  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.000000   
2  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.000000   
3  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.000000   
4  0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.110575   

   1050  1051  1052  1053  1054      1055  1056  1057  1059  106  1060  1061  \
0   0.0   0.0   0.0   0.0   0.0  0.000000   0.0   0.0   0.0  0.0   0.0   0.0   
1   0.0   0.0   0.0   0.0   0.0  0.000000   0.0   0.0   0.0  0.0   0.0   0.0   
2   0.0   0.0   0.0   0.0   0.0  0.011004   0.0   0.0   0.0  0.0   0.0   0.0   
3   0.0   0.0   0.0   0.0   0.0  0.000000   0.0   0.0   0.0  0.0   0.0   0.0   
4   0.0   0.0   0.0   0.0   0.0  0.000000   0.0   0.0   0.0  0.0   0.0   0.0   

   1062  1063  1064  1065  1066  1067  1068  1069  107  1070  1072      1073  \
0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.0   0.0   0.0  0.000000   
1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.0   0.0   0.0  0.000000   
2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.0   0.0   0.0  0.005468   
3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.0   0.0   0.0  0.000000   
4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  0.0   0.0   0.0  0.000000   

   1074  1075  1076  1077  1078  1079       108      1080  1081  1082  1083  \
0   0.0   0.0   0.0   0.0   0.0   0.0  0.000000  0.000000   0.0   0.0   0.0   
1   0.0   0.0   0.0   0.0   0.0   0.0  

In [5]:
# create the machine learing dataframe
df_ml = df.loc[:, ['id', 'judge']]

# add the vectorized vocabulary to the df_ml dataframe
df_ml = pd.concat([df_ml, df_tfidf], axis=1)

In [6]:
# divide the data into label and features for use in ml models
X = df_ml.iloc[:, 2:]
y = df_ml.loc[:, 'judge']

# split the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=6, stratify=y)

### Logistic Regression

In [7]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=6)
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
print('Accuracy on training set = {}'.format(lr.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(lr.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy on training set = 0.6063638315087267
Accuracy on test set = 0.4777095304141074


### Stochastic Gradient Descent Classifier

In [8]:
from sklearn.linear_model import SGDClassifier
sgdc = SGDClassifier(random_state=6)
sgdc.fit(X_train, y_train)
y_pred_sgdc = sgdc.predict(X_test)
print('Accuracy on training set = {}'.format(sgdc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(sgdc.score(X_test, y_test)))

Accuracy on training set = 0.8424398540314724
Accuracy on test set = 0.5780661779274817


### Linear Support Vector Machine Classifier

In [9]:
from sklearn.svm import LinearSVC
svc = LinearSVC(random_state=6)
svc.fit(X_train, y_train)

y_pred_svc = svc.predict(X_test)
print('Accuracy on training set = {}'.format(svc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(svc.score(X_test, y_test)))

Accuracy on training set = 0.9884909430161325
Accuracy on test set = 0.6752526253219735


### Summary

Here is a summary of the results:

| Model               | Train (min_df=0.02) | Test (min_df=0.02) | Train (min_df=0.001) | Test (min_df=0.001) |
|---------------------|---------------------|--------------------|----------------------|---------------------|
| Linear SVM          | 0.9044              | 0.6814             | 0.9885               | 0.6753              |
| Logistic Regression | 0.5816              | 0.4922             | 0.6064               | 0.4778              |
| SGD                 | 0.7290              | 0.5581             | 0.8424               | 0.5781              |

Expanding the vocabulary from 3677 to 16905 words did not meaningfully improve model performance.  The Linear SVM and Logistic Regression models actually performed worse on the test set and the SGD Classifier improved by only 2%.  This isn't necessarily surprising and demonstrates that expanding the vocabulary even further is not likely to improve model performance, as it is adding less-utilized, and likely less-significant, words to the vocabulary.  Expanding the vocabulary further would also open the door for more OCR errors to be included, which we know is an issue with this dataset.

For these reasons, I will proceed with the 3,677 word vocabulary, as it will make model building more efficient without sacrificing much, if any, performance.

### Combined Feature Performance

Next, I will combine the categorical and continuous variables with the tfidf-vectorized vocabulary to see how the models perform with both sets of features.

In [2]:
# load the categorical and continuous variables
df_ml = pd.read_csv('df_ml_1.csv', index_col=0)
df_ml.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80749 entries, 0 to 80748
Columns: 243 entries, id to subjectivity
dtypes: float64(5), int64(237), object(1)
memory usage: 150.3+ MB


In [3]:
# load the tfidf vectorized vocabulary
df_tfidf = pd.read_csv('df_tfidf_02.csv', index_col=0)
df_tfidf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80749 entries, 0 to 80748
Columns: 3677 entries, 00 to your
dtypes: float64(3677)
memory usage: 2.2 GB


In [5]:
# add the vectorized vocabulary to the df_ml dataframe
df_ml = pd.concat([df_ml, df_tfidf], axis=1)
df_ml.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 80749 entries, 0 to 80748
Columns: 3920 entries, id to your
dtypes: float64(3682), int64(237), object(1)
memory usage: 2.4+ GB


In [6]:
df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  year_1779  year_1784  year_1787  year_1789  \
0              1               0          0          0          0          0   
1              1               0          0          0          0          0   
2              1               0          0          0          0          0   
3              1               0          0          0          0          0   
4              1               0          0          0          0          0   

   year_1790  year_1791  year_1792  year_1793  year_1794  year_1795  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1796  year_1797  year_1798  year_1799  year_1800  year_1801  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1802  year_1803  year_1804  year_1805  year_1806  year_1807  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1808  year_1809  year_1810  year_1811  year_1812  year_1813  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1814  year_1815  year_1816  year_1817  year_1818  year_1819  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1820  year_1821  year_1822  year_1823  year_1824  year_1825  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1826  year_1827  year_1828  year_1829  year_1830  year_1831  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          

In [7]:
# divide the data into label and features for use in ml models
X = df_ml.iloc[:, 2:]
y = df_ml.loc[:, 'judge']

# split the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=6, stratify=y)

### Logistic Regression

In [8]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=6)
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
print('Accuracy on training set = {}'.format(lr.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(lr.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Accuracy on training set = 0.7363649873681082
Accuracy on test set = 0.6236378046364177


### Stochastic Gradient Descent Classifier

In [9]:
from sklearn.linear_model import SGDClassifier
sgdc = SGDClassifier(random_state=6)
sgdc.fit(X_train, y_train)
y_pred_sgdc = sgdc.predict(X_test)
print('Accuracy on training set = {}'.format(sgdc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(sgdc.score(X_test, y_test)))

Accuracy on training set = 0.7165007182840442
Accuracy on test set = 0.5903011690112938


### Linear Support Vector Machine Classifier

In [10]:
from sklearn.svm import LinearSVC
svc = LinearSVC(random_state=6)
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)
print('Accuracy on training set = {}'.format(svc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(svc.score(X_test, y_test)))

Accuracy on training set = 0.9654563167715197
Accuracy on test set = 0.7418763621953636


### Summary

Here are the results:

| Model               | Train  | Test   |
|---------------------|--------|--------|
| Linear SVM          | 0.9655 | 0.7419 |
| Logistic Regression | 0.7364 | 0.6236 |
| SGD                 | 0.7165 | 0.5903 |

Combining the features improved the accuracy of the Linear SVM and Logistic Regression models significantly.  Both models are producing favorable accuracy percentages considering that the model has to identify the correct author from a pool of over 150 judges.  The SGD Classifier did not improve as significantly as the other two models and is now producing the lowest accuracy, so I have elected to move forward without it.

## Model Hyperparameter Tuning

Now that I have selected the two best-performing classifiers, I will do some model tuning to see if that further improves their performance.

### Logistic Regression

First, I will try the `newton-cg` solver to see if it will increase model performance over the default `lbfgs` solver:

In [ ]:
from sklearn.linear_model import LogisticRegression
solver_list = ['newton-cg']
for solver in solver_list:
    lr = LogisticRegression(solver=solver, multi_class='auto')
    lr.fit(X_train, y_train)
    y_pred_lr = lr.predict(X_test)
    print('Solver value {} Training Set Accuracy: {}'.format(solver, lr.score(X_train, y_train)))
    print('Solver value {} Test Set Accuracy: {}'.format(solver, lr.score(X_test, y_test)))

Solver value newton-cg Training Set Accuracy: 0.7517214048645168
Solver value newton-cg Test Set Accuracy: 0.6322072518327719


The `newton-cg` solver produced the same accuracy results, so I will stick with the recommended default `lbfgs` solver.  Next, I will tune the C parameter, which controls the regularization strength of the model:

In [8]:
from sklearn.linear_model import LogisticRegression
c_list = [0.01, 0.1, 1, 10, 100, 1000, 10000]
for c in c_list:
    lr = LogisticRegression(C=c, solver='lbfgs', multi_class='auto', max_iter=1000)
    lr.fit(X_train, y_train)
    y_pred_lr = lr.predict(X_test)
    print('C value {} Training Set Accuracy: {}'.format(c, lr.score(X_train, y_train)))
    print('C value {} Test Set Accuracy: {}'.format(c, lr.score(X_test, y_test)))

C value 0.01 Training Set Accuracy: 0.13835636795957795
C value 0.01 Test Set Accuracy: 0.13552605508222706
C value 0.1 Training Set Accuracy: 0.41714965076534405
C value 0.1 Test Set Accuracy: 0.37859124232217156
C value 1 Training Set Accuracy: 0.7516718680338832
C value 1 Test Set Accuracy: 0.6322072518327719
C value 10 Training Set Accuracy: 0.9649609484651839
C value 10 Test Set Accuracy: 0.7231523677432138


C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


C value 100 Training Set Accuracy: 0.9997027790161985
C value 100 Test Set Accuracy: 0.7294927679809788


C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


C value 1000 Training Set Accuracy: 0.9999009263387328
C value 1000 Test Set Accuracy: 0.7243907271646522


C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


C value 10000 Training Set Accuracy: 0.9999009263387328
C value 10000 Test Set Accuracy: 0.720229839508619


Here is a summary of the results of our hyperparameter tuning for the Logistic Regression classifier:

| Model               | Solver    | C Value | Train  | Test   |
|---------------------|-----------|---------|--------|--------|
| Logistic Regression | lbfgs     | 100     | 0.9998 | 0.7295 |
| Logistic Regression | lbfgs     | 1000    | 0.9999 | 0.7244 |
| Logistic Regression | lbfgs     | 10      | 0.9650 | 0.7232 |
| Logistic Regression | lbfgs     | 10000   | 0.9999 | 0.7202 |
| Logistic Regression | lbfgs     | 1       | 0.7517 | 0.6322 |
| Logistic Regression | newton-cg | 1       | 0.7517 | 0.6322 |
| Logistic Regression | lbfgs     | 0.1     | 0.4171 | 0.3786 |
| Logistic Regression | lbfgs     | 0.001   | 0.1384 | 0.1355 |

The tuning suggests that a C parameter value of 100 will result in the marginally best performance of the model.  I also will need to increase the number of iterations to higher than 1000 to allow the model to converge using that C parameter.

### Linear Support Vector Machine Classifier

For the Linear Support Vector Machine Classifier, I will tune the C parameter:

In [11]:
from sklearn.svm import LinearSVC
c_list = [0.01, 0.1, 1, 10, 100, 1000]
for c in c_list:
    svc = LinearSVC(C=c, max_iter=1500, random_state=6)
    svc.fit(X_train, y_train)
    y_pred_svc = svc.predict(X_test)
    print('C value {} Training Set Accuracy: {}'.format(c, svc.score(X_train, y_train)))
    print('C value {} Test Set Accuracy: {}'.format(c, svc.score(X_test, y_test)))

C value 0.01 Training Set Accuracy: 0.5204999917438615
C value 0.01 Test Set Accuracy: 0.456905092133941
C value 0.1 Training Set Accuracy: 0.8104885982728158
C value 0.1 Test Set Accuracy: 0.6641569249058847
C value 1 Training Set Accuracy: 0.9654563167715197
C value 1 Test Set Accuracy: 0.7418763621953636
C value 10 Training Set Accuracy: 0.9975561830220769
C value 10 Test Set Accuracy: 0.7231523677432138


C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\svm\base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


C value 100 Training Set Accuracy: 0.9995706808011757
C value 100 Test Set Accuracy: 0.6945214979195562


C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\svm\base.py:929: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)


C value 1000 Training Set Accuracy: 0.9995871930780535
C value 1000 Test Set Accuracy: 0.6842678819100456


It appears that the default C parameter value of 1 produced the highest accuracy for the Linear SVC model.

### Summary of Hyperparameter Tuning

Here is a summary of the results of the hyperparameter tuning:

| Model               | Solver    | C Value | Train  | Test   |
|---------------------|-----------|---------|--------|--------|
| Linear SVC          |           | 1       | 0.9655 | 0.7419 |
| Logistic Regression | lbfgs     | 100     | 0.9998 | 0.7295 |
| Logistic Regression | lbfgs     | 1000    | 0.9999 | 0.7244 |
| Linear SVC          |           | 10      | 0.9976 | 0.7232 |
| Logistic Regression | lbfgs     | 10      | 0.9650 | 0.7232 |
| Logistic Regression | lbfgs     | 10000   | 0.9999 | 0.7202 |
| Linear SVC          |           | 100     | 0.9996 | 0.6945 |
| Linear SVC          |           | 1000    | 0.9996 | 0.6843 |
| Linear SVC          |           | 0.1     | 0.8105 | 0.6642 |
| Logistic Regression | lbfgs     | 1       | 0.7517 | 0.6322 |
| Logistic Regression | newton-cg | 1       | 0.7517 | 0.6322 |
| Linear SVC          |           | 0.01    | 0.5205 | 0.4569 |
| Logistic Regression | lbfgs     | 0.1     | 0.4171 | 0.3786 |
| Logistic Regression | lbfgs     | 0.001   | 0.1384 | 0.1355 |

After tuning, the best Linear SVC model is generating an accuracy percentage of 74.19% on the test set and the best Logistic Regression Classifier is generating an accuracy percentage of 72.95%.  Both models are performing well at this point, especially considering that there are approximately 150 judges who authored opinions in the dataset.

As a final step, I will add some `SpaCy` word embeddings vectors for each opinion to see if they improve model performance.

### SpaCy Vectors

The `SpaCy` package converts documents into 300 dimension vectors that attempt to capture the substance of the text with numerical values.  I have previously created these vectors and saved them to a local file, so I will import that file and then add the vectors as additional features to the `df_ml` dataframe.  I have provided to code I used to create the vectors in the next cell (as comments because I do not need to run the code at this time).

In [ ]:
# import and instantiate SpaCy
# import spacy
# nlp = spacy.load('en_core_web_md')

# create spacy vectors
# from tqdm import tqdm
# for row in tqdm(range(len(df_spacy))):
    #doc = nlp(df.loc[row, 'opinion'])
    #for i in range(300):
        #df_spacy.loc[row, 'spacy_{}'.format(i)] = doc.vector[i]

In [4]:
# import file
df_spacy = pd.read_csv('df_spacy.csv', index_col=0)

In [7]:
df_spacy.head()

spacy_0   spacy_1   spacy_2   spacy_3   spacy_4   spacy_5   spacy_6  \
0 -0.018235  0.141126 -0.077766 -0.041140 -0.001164 -0.014620  0.032196   
1 -0.084114  0.211868 -0.025314 -0.079775 -0.034274 -0.037462  0.057598   
2 -0.031218  0.177579 -0.110459 -0.015997  0.022760 -0.023197 -0.004636   
3 -0.021508  0.177556 -0.105615 -0.043884  0.002879 -0.031908 -0.003519   
4  0.017297  0.159210 -0.056437 -0.064930  0.023885 -0.034701 -0.000850   

    spacy_7   spacy_8   spacy_9  spacy_10  spacy_11  spacy_12  spacy_13  \
0 -0.082413  0.024770  2.116468 -0.192804  0.049729  0.141365 -0.016655   
1 -0.048959  0.001335  2.026258 -0.278814  0.090320  0.200667 -0.007330   
2 -0.089868 -0.021976  2.266324 -0.141792  0.060227  0.134374 -0.010603   
3 -0.060925 -0.005755  2.215132 -0.139976  0.065458  0.137746 -0.002435   
4 -0.021355  0.030055  2.110572 -0.157416  0.069915  0.146728 -0.033003   

   spacy_14  spacy_15  spacy_16  spacy_17  spacy_18  spacy_19  spacy_20  \
0 -0.169243 -0.062276 -0.033327  1.070306 -0.115193 -0.033005  0.031212   
1 -0.113698 -0.018429 -0.072254  0.934032 -0.140462  0.001633  0.021376   
2 -0.211327 -0.059041 -0.051099  0.932276 -0.130797 -0.049704  0.020150   
3 -0.207878 -0.073383 -0.055856  0.981315 -0.115957 -0.038079 -0.014753   
4 -0.175534 -0.064603 -0.044850  1.090652 -0.106103 -0.032908  0.012718   

   spacy_21  spacy_22  spacy_23  spacy_24  spacy_25  spacy_26  spacy_27  \
0 -0.071661 -0.031958 -0.008337  0.065750  0.043408 -0.052442 -0.077462   
1 -0.091945 -0.027910 -0.026018  0.147895  0.093733 -0.105549  0.026241   
2 -0.078448 -0.034943  0.004263  0.073384  0.021532 -0.054368 -0.055199   
3 -0.104550 -0.028625  0.008731  0.059671  0.007517 -0.041222 -0.067193   
4 -0.080671 -0.024293  0.006308  0.020825  0.027461 -0.085772 -0.059748   

   spacy_28  spacy_29  spacy_30  spacy_31  spacy_32  spacy_33  spacy_34  \
0 -0.011728 -0.059205 -0.040694  0.066518 -0.024291  0.079856  0.069881   
1  0.033703 -0.025820 -0.028645 -0.020044 -0.047317  0.083274  0.011339   
2  0.000638 -0.090853 -0.036661  0.074017 -0.053181  0.084578  0.053516   
3  0.021957 -0.082004 -0.056516  0.064191 -0.064978  0.065994  0.048644   
4 -0.065279 -0.025525 -0.095215  0.093233 -0.034150  0.087134  0.057596   

   spacy_35  spacy_36  spacy_37  spacy_38  spacy_39  spacy_40  spacy_41  \
0 -0.082784 -0.011462 -0.000150 -0.071227 -0.070347 -0.024128  0.029422   
1 -0.069034  0.045600 -0.070799 -0.092307 -0.077901 -0.038027  0.102327   
2 -0.104696 -0.019632  0.033720 -0.078516 -0.075380  0.012850  0.014253   
3 -0.093732  0.000837  0.003621 -0.079783 -0.069343 -0.020788  0.009027   
4 -0.064533 -0.012003  0.007128 -0.093714 -0.067563 -0.056920  0.051694   

   spacy_42  spacy_43  spacy_44  spacy_45  spacy_46  spacy_47  spacy_48  \
0  0.077631 -0.088816 -0.031311 -0.014032 -0.116316  0.063228 -0.001043   
1  0.140531 -0.113886 -0.010942 -0.007901 -0.245600  0.067280 -0.027769   
2  0.082791 -0.115984 -0.003837 -0.000222 -0.121734 -0.011534  0.041311   
3  0.093332 -0.105704 -0.015653 -0.016875 -0.121958  0.023353  0.019103   
4  0.097947 -0.067258  0.007046 -0.003752 -0.070797  0.017701  0.006285   

   spacy_49  spacy_50  spacy_51  spacy_52  spacy_53  spacy_54  spacy_55  \
0  0.004374 -0.039147  0.051738 -0.047044  0.039755 -0.034598  0.012541   
1 -0.072281 -0.010517  0.017179  0.090949 -0.052786  0.036153 -0.022454   
2  0.031829 -0.060940  0.064460 -0.076666  0.046999  0.005713  0.022953   
3  0.024854 -0.072786  0.072361 -0.057798  0.044385  0.005329  0.039805   
4  0.037754 -0.030818  0.042224 -0.068200  0.047317 -0.013456 -0.002436   

   spacy_56  spacy_57  spacy_58  spacy_59  spacy_60  spacy_61  spacy_62  \
0  0.026246 -0.030908  0.002282 -0.062070 -0.011083 -0.026285  0.018029   
1 -0.057134 -0.037015 -0.032210 -0.128844 -0.121354 -0.028959  0.098923   
2 -0.013933 -0.049501 -0.015271 -0.072222 -0.025141 -0.024457  0.016111   
3  0.003755 -0.049763 -0.014534 -0.065294 -0.028938 -0.039785  0.019084   
4  0.051357 -0.026557

In [6]:
del df_spacy['id']
del df_spacy['judge']

In [8]:
df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  year_1779  year_1784  year_1787  year_1789  \
0              1               0          0          0          0          0   
1              1               0          0          0          0          0   
2              1               0          0          0          0          0   
3              1               0          0          0          0          0   
4              1               0          0          0          0          0   

   year_1790  year_1791  year_1792  year_1793  year_1794  year_1795  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1796  year_1797  year_1798  year_1799  year_1800  year_1801  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1802  year_1803  year_1804  year_1805  year_1806  year_1807  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1808  year_1809  year_1810  year_1811  year_1812  year_1813  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1814  year_1815  year_1816  year_1817  year_1818  year_1819  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1820  year_1821  year_1822  year_1823  year_1824  year_1825  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1826  year_1827  year_1828  year_1829  year_1830  year_1831  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          

In [9]:
# add the spacy word embedded vectors to the df_ml dataframe
df_ml = pd.concat([df_ml, df_tfidf, df_spacy], axis=1)
df_ml.head()

id         judge  type_concurrence  \
0  53839          frye                 0   
1  53841    per_curiam                 0   
2  53843  parker_sarah                 0   
3  53847      whichard                 0   
4  53848          webb                 0   

   type_concurring-in-part-and-dissenting-in-part  type_dissent  \
0                                               0             0   
1                                               0             0   
2                                               0             0   
3                                               0             0   
4                                               0             0   

   type_majority  type_rehearing  year_1779  year_1784  year_1787  year_1789  \
0              1               0          0          0          0          0   
1              1               0          0          0          0          0   
2              1               0          0          0          0          0   
3              1               0          0          0          0          0   
4              1               0          0          0          0          0   

   year_1790  year_1791  year_1792  year_1793  year_1794  year_1795  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1796  year_1797  year_1798  year_1799  year_1800  year_1801  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1802  year_1803  year_1804  year_1805  year_1806  year_1807  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1808  year_1809  year_1810  year_1811  year_1812  year_1813  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1814  year_1815  year_1816  year_1817  year_1818  year_1819  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1820  year_1821  year_1822  year_1823  year_1824  year_1825  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          0          0          0          0          0          0   
4          0          0          0          0          0          0   

   year_1826  year_1827  year_1828  year_1829  year_1830  year_1831  \
0          0          0          0          0          0          0   
1          0          0          0          0          0          0   
2          0          0          0          0          0          0   
3          

In [10]:
# divide the data into label and features for use in ml models
X = df_ml.iloc[:, 2:]
y = df_ml.loc[:, 'judge']

# split the data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=6, stratify=y)

### Logistic Regression

In [11]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(C=100, solver='lbfgs', multi_class='auto', max_iter=3000, random_state=6)
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)
print('Accuracy on training set = {}'.format(lr.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(lr.score(X_test, y_test)))

C:\Users\jacob\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:947: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)


Accuracy on training set = 0.9999339508924886
Accuracy on test set = 0.7542599564097484


### Linear Support Vector Machine Classifier

In [12]:
from sklearn.svm import LinearSVC
svc = LinearSVC(max_iter=1500, random_state=6)
svc.fit(X_train, y_train)
y_pred_svc = svc.predict(X_test)
print('Accuracy on training set = {}'.format(svc.score(X_train, y_train)))
print('Accuracy on test set = {}'.format(svc.score(X_test, y_test)))

Accuracy on training set = 0.9746866795462427
Accuracy on test set = 0.7675847037844263


The `SpaCy` vectors improved the accuracy of both models by approximately 2%, demonstrating that the substance of the opinions provides signal for the models.  Here are the final results:

| Model               | Train  | Test   |
|---------------------|--------|--------|
| Linear SVC          | 0.9747 | 0.7676 |
| Logistic Regression | 0.9999 | 0.7542 |

With both models generating an accuracy percentage above 75%, I have succeeded in building two models that will provide value to my clients.